subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c

```diff
@@ -0,0 +1,455 @@
+	vr->shm_size = shm_size(num_desc) - VDEV_STATUS_SIZE;
+
+	vr->rx_addr = vr->shm_addr + VRING_COUNT * vq_ring_size(num_desc);
+	vr->tx_addr = vr->rx_addr + vring_size(num_desc, VRING_ALIGNMENT);
+
+	vr->status_reg_addr = conf->shm_addr;
+
+	vr->vring_size = num_desc;
+}
+
+static void virtio_notify_cb(struct virtqueue *vq, void *priv)
+{
+	struct backend_config_t *conf = priv;
+
+	if (conf) {
+		mbox_send(&conf->mbox_tx, NULL);
+	}
+}
+
+static void mbox_callback_process(struct k_work *item)
+{
+	struct backend_data_t *data;
+	unsigned int vq_id;
+
+	data = CONTAINER_OF(item, struct backend_data_t, mbox_work);
+	vq_id = (data->role == VIRTIO_DEV_MASTER) ? VIRTQUEUE_ID_HOST : VIRTQUEUE_ID_REMOTE;
+
+	virtqueue_notification(data->vr.vq[vq_id]);
+}
+
+static void mbox_callback(const struct device *instance, uint32_t channel,
+			  void *user_data, struct mbox_msg *msg_data)
+{
+	struct backend_data_t *data = user_data;
+
+	k_work_submit_to_queue(&mbox_wq, &data->mbox_work);
+}
+
+static int mbox_init(const struct device *instance)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *data = instance->data;
+	int err;
+
+	k_work_queue_start(&mbox_wq, mbox_stack, K_KERNEL_STACK_SIZEOF(mbox_stack),
+			   WQ_PRIORITY, NULL);
+
+	k_work_init(&data->mbox_work, mbox_callback_process);
+
+	err = mbox_register_callback(&conf->mbox_rx, mbox_callback, data);
+	if (err != 0) {
+		return err;
+	}
+
+	return mbox_set_enabled(&conf->mbox_rx, 1);
+}
+
+static int register_ept(const struct device *instance, void **token,
+			const struct ipc_ept_cfg *cfg)
+{
+	struct backend_data_t *data = instance->data;
+	struct ipc_rpmsg_instance *rpmsg_inst;
+	struct ipc_rpmsg_ept *rpmsg_ept;
+	bool ept_cached;
+	int err;
+
+	/* Instance is still being initialized */
+	if (data->state == STATE_BUSY) {
+		return -EBUSY;
+	}
+
+	/* Instance is not initialized */
+	if (data->state == STATE_READY) {
+		return -EINVAL;
+	}
+
+	/* Empty name is not valid */
+	if (cfg->name == NULL || cfg->name[0] == '\0') {
+		return -EINVAL;
+	}
+
+	rpmsg_inst = &data->rpmsg_inst;
+
+	k_mutex_lock(&rpmsg_inst->mtx, K_FOREVER);
+	ept_cached = get_ept(rpmsg_inst, &rpmsg_ept, cfg->name);
+
+	if (rpmsg_ept == NULL) {
+		k_mutex_unlock(&rpmsg_inst->mtx);
+		return -ENOENT;
+	}
+
+	rpmsg_ept->cb = &cfg->cb;
+	rpmsg_ept->priv = cfg->priv;
+	rpmsg_ept->bound = false;
+	rpmsg_ept->ep.priv = rpmsg_ept;
+
+	if (ept_cached) {
+		/*
+		 * The endpoint was cached in the NS bind callback. We can finally
+		 * advertise it.
+		 */
+		k_mutex_unlock(&rpmsg_inst->mtx);
+		advertise_ept(rpmsg_inst, rpmsg_ept, cfg->name, rpmsg_ept->dest);
+	} else {
+		/*
+		 * There is no endpoint in the cache because the REMOTE has
+		 * not registered the endpoint yet. Cache it.
+		 */
+		strncpy(rpmsg_ept->name, cfg->name ? cfg->name : "", sizeof(rpmsg_ept->name));
+		k_mutex_unlock(&rpmsg_inst->mtx);
+	}
+
+	err = ipc_rpmsg_register_ept(rpmsg_inst, data->role, rpmsg_ept);
```

Comment:
I'm thinking what happens if `register_ept()` is called in host first and right after on remote, so that remote receives `ns_bind_cb()` while the mutex is locked.

Remote code would see that `ept_cached == false`, it would cache the endpoint and unlock the mutex. Unlocking the mutex would trigger `ns_bind_cb()` handler which would call `rpmsg_create_ept()`, because the endpoint would be already cached. And right after that, `ipc_rpmsg_register_ept()` is called, which also calls `rpmsg_create_ept()` for the same endpoint.

Is it OK to call `rpmsg_create_ept()` twice for the same endpoint? What if both `register_ept()` and `ns_bind_cb()` are executed from threads with the same priority? Is `rpmsg_create_ept()` reentrant?

Reply:
> remote receives ns_bind_cb() while the mutex is locked.

This is not going to happen because `ns_bind_cb()` is only called on host (see huge comment in `subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.h`). 

More in detail the `ns_bind_cb()` callback is called **only** on the host side when on the remote side `rpmsg_create_ept()` is called with destination `RPMSG_ADDR_ANY` in  https://github.com/zephyrproject-rtos/zephyr/blob/8fbbf2baefadbf4a3d08e19a7f60de1b989dc36e/subsys/ipc/ipc_service/lib/ipc_rpmsg.c#L54-L57 during the endpoint registration.

This call is triggering the `ns_bind_cb()` on the host side. On the host side we call indeed `rpmsg_create_ept()` in https://github.com/zephyrproject-rtos/zephyr/blob/32812e7b62e1fb88c471f773332a8f274b115e6e/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L121-L122 but this time we are specifying the destination address of the remote endpoint so this is not triggering anything on the remote side (see https://github.com/OpenAMP/open-amp/blob/master/lib/rpmsg/rpmsg.c#L301-L303)

Reply:
So, it means that `if (ept_cached)` from line 317 can be true only in the host role, right? Maybe we could add an assertion in there to check the role. It would help following the code.

I think I finally understood how the code flow works here. I'm not sure how to improve readability. Maybe comment for `ns_bind_cb()` and `advertise_ept()` that these functions are called only in the host role?